nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/test/java/org/apache/nifi/json/TestJsonTreeRowRecordReader.java

```diff
@@ -1128,29 +1140,82 @@ void testStartFromNestedFieldThenStartObject() throws IOException, MalformedReco
+        ));
+
+        RecordSchema recordSchema = new SimpleRecordSchema(Collections.singletonList(
+                new RecordField("account", RecordFieldType.RECORD.getRecordDataType(accountSchema))
+        ));
+
+        RecordSchema expectedRecordSchema = accountSchema;
+
+        List<Object> expected = Collections.singletonList(
+                new MapRecord(expectedRecordSchema, new HashMap<String, Object>() {{
+                    put("id", 42);
+                    put("balance", 4750.89);
+                }})
+        );
+
+        testReadRecords(jsonPath, recordSchema, expected, StartingFieldStrategy.NESTED_FIELD,
+                "account", SchemaApplicationStrategy.WHOLE_JSON);
+    }
+
+    @Test
+    void testStartFromNestedArrayWithWholeJsonSchemaScope() throws IOException, MalformedRecordException {
+        String jsonPath = "src/test/resources/json/single-element-nested-array.json";
+
+        RecordSchema accountSchema = new SimpleRecordSchema(Arrays.asList(
+                new RecordField("id", RecordFieldType.INT.getDataType()),
+                new RecordField("balance", RecordFieldType.DOUBLE.getDataType())
+        ));
+
+        RecordSchema recordSchema = new SimpleRecordSchema(Collections.singletonList(
+                new RecordField("accounts", RecordFieldType.ARRAY.getArrayDataType(RecordFieldType.RECORD.getRecordDataType(accountSchema)))
+        ));
+
+        RecordSchema expectedRecordSchema = accountSchema;
+
+        List<Object> expected = Arrays.asList(
+                new MapRecord(expectedRecordSchema, new HashMap<String, Object>() {{
+                    put("id", 42);
+                    put("balance", 4750.89);
+                }}),
+                new MapRecord(expectedRecordSchema, new HashMap<String, Object>() {{
+                    put("id", 43);
+                    put("balance", 48212.38);
+                }})
+        );
+
+        testReadRecords(jsonPath, recordSchema, expected, StartingFieldStrategy.NESTED_FIELD,
+                "accounts", SchemaApplicationStrategy.WHOLE_JSON);
     }
 
     private void testReadRecords(String jsonPath, List<Object> expected) throws IOException, MalformedRecordException {
-        // GIVEN
         final File jsonFile = new File(jsonPath);
-
         try (
             InputStream jsonStream = new ByteArrayInputStream(FileUtils.readFileToByteArray(jsonFile))
         ) {
             RecordSchema schema = inferSchema(jsonStream, StartingFieldStrategy.ROOT_NODE, null);
-
-            // WHEN
-            // THEN
             testReadRecords(jsonStream, schema, expected);
         }
     }
 
-    private void testReadRecords(String jsonPath, List<Object> expected, StartingFieldStrategy strategy, String startingFieldName) throws IOException, MalformedRecordException {
+    private void testReadRecords(String jsonPath, List<Object> expected, StartingFieldStrategy strategy,
```

Comment:
This method now accepts a `SchemaApplicationStrategy` but always tests the  production code with an inferred schema that is based on the `staringFieldName` i.e. a nested schema.

In other words if `SchemaApplicationStrategy.WHOLE_JSON` is provided the test case does not correspond to any possible real-life scenario.

I would either remove the `schemaApplicationStrategy` parameter (and consider this utility method to be restricted to cases when the nested schema is to be used a.k.a. `SchemaApplicationStrategy` is `SELECTED_PART`) or update the `inferSchema` logic. I'd vote for the former - in my opinion such logic shouldn't be used in the test code in the first place.